### ISSUE TITLE
Feature request: Being able to retrieve swarm node name / host's hostname

### ISSUE THREAD

**Issue by mostolog** [moby/moby#30966]
As requested in https://forums.docker.com/t/how-to-specify-our-own-unique-node-id-or-node-name-to-docker-swarm-nodes/28109 it could be interesting to be able to make node names more friendly, having a **PlaceHolder {{.Node.Name}}** with the host's hostname in https://docs.docker.com/engine/reference/commandline/service_create/#/create-services-using-templates

Regards

**Comment by thaJeztah** [comment]
ping @nishanttotla @aaronlehmann PTAL, I think this came up recently (hostname vs name)

**Comment by nishanttotla** [comment]
@mostolog If I understand correctly, are you asking for the ability to set Swarm node names separately from the machine hostname?

Currently, Docker just picks up the machine hostname updates the Swarm node name to the hostname. SwarmKit already supports this capability, so it would be as simple as adding a flag to `docker swarm create` or `docker swarm join`. I can work on it if we want something like that.

**Comment by mostolog** [comment]
As you said, swarm nodes get host's hostname as swarm node name (internally). That's perfectly correct for me.

After deploying some services with default setting, _ps_ shows something like:
```
37bbf2b56dbf serviceBar.1.ngneetjtxwhm2eijfs1gvt60t
a90e2d59cd77 serviceBar.2.d4lf8h3gqj9e7gyhfi3lgrtdt
ed6d150d7c84 serviceBar.3.iz85bp0sc5t6e3b15w6gtteus
8d7343dfea35 serviceFoo.1.xe9i3vzatpntmmpft0cjup2wa
16a283d6e54e serviceFoo.2.4ofjgosnl44p63rkyrlhe7ut0
89f5d450c578 serviceFoo.3.vn1c62qcmagv4m4u9gzgcdj1k
5783bbdbf238 serviceEtc.1.xe9i3vzatpnt2eijfs1gvt60t
...
```
I don't like that ugly digest...but I could deal with it.
Ok...those containers have ugly names, and if I connect to them on different terminals I wouldn't know where am I. So I decided to use --hostname flag.
> --hostname: {{.Service.Name}}-{{.Task.Slot}}

That's ok, but I would love to know in which node is that container running, so
> --hostname: {{.Service.Name}}-{{.Task.Slot}}-{{.Node.Name}}

and  _ps_ would display:
```
serviceBar-1-node1 serviceBar.1
serviceBar-2-node3 serviceBar.2
serviceBar-3-node1 serviceBar.3
serviceFoo-1-node2 serviceFoo.1
serviceFoo-2-node3 serviceFoo.2
serviceFoo-3-node1 serviceFoo.3
serviceEtc-1-node2 serviceEtc.1
...
```

and connecting to that container would have set $PS1 and hostname to it (eg:serviceFoo-3-node1)
To sum up: now I know where each container is running and where am I if connected to any container
:)

**Comment by mostolog** [comment]
https://github.com/moby/moby/issues/32561

**Comment by mostolog** [comment]
https://github.com/docker/swarmkit/issues/1951

**Comment by thaJeztah** [comment]
This works on master, and will be included in the upcoming Docker 17.10 release;

```bash
$ hostname testing123

$ docker service create \
--name foo \
--hostname "{{.Service.Name}}-{{.Task.Slot}}-{{.Node.Hostname}}" \
nginx:alpine

$ docker inspect foo.1.k51r3eclvcelsxy6jthtavkwa --format '{{.Config.Hostname}}'
foo-1-testing123
```

Also see https://github.com/docker/swarmkit/issues/1951, and https://github.com/moby/moby/pull/34686
